src/service/url-replacements-impl.js

```diff
@@ -408,55 +483,161 @@ export class UrlReplacements {
+   * asynchronous resolver where expand will use async and expandSync the sync
+   * version.
+   * @param {string} varName
+   * @param {!SyncResolverDef} syncResolver
+   * @return {!UrlReplacements}
+   * @private
+   */
+  set_(varName, syncResolver) {
+    dev.assert(varName.indexOf('RETURN') == -1);
+    this.replacements_[varName] =
+        this.replacements_[varName] || {sync: undefined, async: undefined};
+    this.replacements_[varName].sync = syncResolver;
+    this.replacementExpr_ = undefined;
+    return this;
+  }
+
+  /**
+  * Sets an async value resolver for the variable with the specified name.
+  * The value resolver may optionally take an extra parameter.
+  * Can be called in conjuction with setAsync to allow for additional
+  * asynchronous resolver where expand will use async and expandSync the sync
+  * version.
    * @param {string} varName
-   * @param {function(*):*} resolver
+   * @param {!AsyncResolverDef} asyncResolver
    * @return {!UrlReplacements}
    * @private
    */
-  set_(varName, resolver) {
+  setAsync_(varName, asyncResolver) {
     dev.assert(varName.indexOf('RETURN') == -1);
-    this.replacements_[varName] = resolver;
+    this.replacements_[varName] =
+        this.replacements_[varName] || {sync: undefined, async: undefined};
+    this.replacements_[varName].async = asyncResolver;
     this.replacementExpr_ = undefined;
     return this;
   }
 
   /**
+   * Utility function for setting resolver for timing data that supports
+   * sync and async.
+   * @param {string} varName
+   * @param {string} startEvent
+   * @param {string=} endEvent
+   * @return {!UrlReplacements}
+   * @private
+   */
+  setTimingResolver_(varName, startEvent, endEvent) {
+    return this.set_(varName, () => {
+        return this.getTimingDataSync_(startEvent, endEvent);
+      }).setAsync_(varName, () => {
+        return this.getTimingDataAsync_(startEvent, endEvent);
+      });
+  }
+
+  /**
+   * Calling encodeURIComponent with non-string type fails type check despite
+   * browser supporting it (simply converts to string).  This allows for being
+   * explicit about the behavior.
+   * @param {*} val
+   * @return {!string}
+   * @private
+   */
+  encodeValue_(val) {
+    switch (typeof val) {
+      case 'string':
+        return encodeURIComponent(val);
+      case 'number':
+      case 'boolean':
+        return String(val);
+      default:  // null evaluates as object
+        return '';
+    }
+  }
+
+  /**
+   * Synchronously expands the provided URL by replacing all known variables with
+   * their resolved values. Optional `opt_bindings` can be used to add new
+   * variables or override existing ones.  Any async bindings are ignored.
+   * @param {string} url
+   * @param {!Object<string, (ResolverReturnDef|!SyncResolverDef)>=} opt_bindings
+   * @param {!Object<string, ResolverReturnDef>=} opt_collectVars
+   * @return {string}
+   * @private
+   */
+   expandSync(url, opt_bindings, opt_collectVars) {
```

Comment:
Also, re https://github.com/ampproject/amphtml/pull/4153#discussion_r71756858, how can we un-dupe the code between sync and async expansion? Right now they share the majority of the implementation.